Add rule for sol file issue #272

hubris opened the pull request:
The rule sol2go allows to generate a go file from a sol source.
In order to have the correct import paths, abigen must have ethereum-go
repository in its GOPATH that's why some bash hack is needed.

The solc compiler needs to be in PATH.
The rule sol2go will generate a go file from the given sol. Then a library can be built with the standard go_library rule. For example:

sol2go(
  name = "contract_go",
  src = "sharding_manager.sol",
  pkg = "smc",
)

This generate sharding_manager.go and the library can be created wilth:

go_library(
    name = "go_default_library",
     srcs = ["sharding_manager.go"],
     importpath = "github.com/prysmaticlabs/prysm/contracts/sharding-manager-contract",
     visibility = ["//client:__subpackages__"],
     deps = [
         "@com_github_ethereum_go_ethereum//:go_default_library",
         "@com_github_ethereum_go_ethereum//accounts/abi:go_default_library",
         "@com_github_ethereum_go_ethereum//accounts/abi/bind:go_default_library",
         "@com_github_ethereum_go_ethereum//common:go_default_library",
         "@com_github_ethereum_go_ethereum//core/types:go_default_library",
         "@com_github_ethereum_go_ethereum//event:go_default_library",
     ],
)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/prysmaticlabs/prysm?pullRequest=411) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/prysmaticlabs/prysm?pullRequest=411) before we can accept your contribution.<br/><hr/>**no** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/prysmaticlabs/prysm?pullRequest=411) it.</sub>

prestonvanloon commented:
Hey @hubris, I'm glad to see you are making progress.

This is a great start but doesn't quite fit the requirements. One of the primary motivators to use Bazel is that it does not have any implicit dependencies. As long as the user has bazel installed, they should be able to build every target in our project without installing a dependency to their system.

Have you taken a look at using the [go_path rule](https://github.com/bazelbuild/rules_go/blob/27b444a6025b8d0eddff27b2fa170fdc9c93b1af/go/core.rst#go-path) to generate the path structure you need to discover imports via goimports?

Have you looked at adding solc as a bazel toolchain?

hubris commented:
@prestonvanloon Hi, 

- I didn't know the go_path rule, I will try and see if it does the job, thanks.
- For the solc compiler, does it mean we need to create a bazel build of the solidity compiler like it has been done for go-ethereum? It seems to have a lot of dependencies which may make things hard but I can I have deeper look. Unless you know some bazel ready version of solc?

hubris commented:
@prestonvanloon  Does it make sense to use https://github.com/ethereum/solidity/releases as external deps and use the right binary according to the host?

hubris commented:
@prestonvanloon Hi, I have updated my branch. The solc compiler is now downloaded. I added this rule in the WORKSPACE:
```bzl
solc_fetch(
    name = "solc_sdk",
    sdks = {
        "linux_amd64":   ("https://github.com/ethereum/solidity/releases/download/v0.4.23/solc-static-linux",
            "c648d299bd21f4c74ce17954706cfc5a7a9fbbd1e8739546ed8754cacd8ca7c2")
    },
)
```
I tried the go_path rule on abigen, but it is a failure. Just doing this as a test:
```bzl
go_path(
    name = "go_ethereum_path",
    mode = "copy",
    deps = ["@com_github_ethereum_go_ethereum//cmd/abigen"],
)
```
Will fail with an error I don't understand:
```
174, in _add_manifest_entry
                fail("{}: references multiple files (...))
src/github.com/ethereum/go-ethereum/crypto/secp256k1/secp256k1.h: references multiple files (external/com_github_ethereum_go_ethereum/crypto/secp256k1/libsecp256k1/include/secp256k1.h and external/com_github_ethereum_go_ethereum/crypto/secp256k1/libsecp256k1/src/secp256k1.h)
ERROR: Analysis of target '//contracts:go_ethereum_path' failed; build aborted: Analysis of target '//contracts:go_ethereum_path' failed; build aborted
```
So I stayed on my original approach.

hubris commented:
@prestonvanloon Hi, Can you have a look the PR please?

prestonvanloon commented:
Hey @hubris,

Very sorry for the delay. This is great. I think this meets all of the requirements. 

I just have one request to use the generated targets rather than the generated files. Since that may not be very clear, I've submitted a PR for your PR. https://github.com/hubris/prysm/pull/1

Please merge that to your branch, update this branch, then we can merge this and close out the issue and award the bounty.

Thanks so much for your effort!

hubris commented:
@prestonvanloon Hi, ok its done.

prestonvanloon commented:
I this LGTM, I’m afraid it won’t build on Mac or windows though since solc is only provided for Linux amd64

@terenc3t can you confirm this is ok on Mac?

prestonvanloon commented on WORKSPACE:
```diff
@@ -83,6 +83,20 @@ go_repository(
+solc_fetch(
+    name = "solc_sdk",
+    sdks = {
+        "linux_amd64": (
```
Can you add all available OS?

terenc3t replied:
Most mac users use homebrew to install solc https://github.com/ethereum/go-ethereum/wiki/Installation-Instructions-for-Mac
Not sure if that's something we can do here

prestonvanloon replied:
Would the same binary for linux work on mac? How do they build for mac? What is provided by homebrew?

prestonvanloon replied:
To answer your question @hubris, I prefer the first of your three suggestions. 

If we can't find compiled binaries online, let's build it and check it into the repo. I'm fine with not supporting 32 bit systems for now, but we need to support darwin and windows at a minimum.

rauljordan replied:
hey @hubris any update on this?

hubris replied:
@rauljordan Hi, I will work on this tomorrow.

hubris replied:
@prestonvanloon Hi, I tried to add the win64 compiler but I'm getting in troubles just for building prysm on windows.
I downloaded bazel for windows, cloned the repo and from the master, I did ``bazel test //...``. I got:
```bash
INFO: Build options have changed, discarding analysis cache.
ERROR: D:/devel/gopath/src/github.com/hubris/prysm/client/BUILD.bazel:21:1: every rule of type app_layer implicitly depends upon the target '@gzip//:gzip', but this target could not be found because of: no such package '@gzip//': expected value of type 'string or Label or path' for parameter 'from', in method call symlink(NoneType, string) of 'repository_ctx'
ERROR: Analysis of target '//client:image' failed; build aborted: no such package '@gzip//': expected value of type 'string or Label or path' for parameter 'from', in method call symlink(NoneType, string) of 'repository_ctx'
INFO: Elapsed time: 2.209s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (47 packages loaded)
FAILED: Build did NOT complete successfully (47 packages loaded)
````

Did I miss something?

prestonvanloon replied:
@hubris I am not sure that the docker image will build on windows.

You should be able to build the specific target binary. `bazel build //beacon-chain`
Are you synced with master? I thought we add the manual tag to that target so it wouldn't build when you build `//...`.

hubris replied:
@prestonvanloon Thank you for the answer. I tried with latest master (42d93812ced4c5a219371f141fe80ec07a658c52).
Doing ``bazel build //beacon-chain`` in cmd.exe still give me errors (doing the same in bash gives similar errors):
```
DEBUG: C:/users/hubris/_bazel_hubris/gcsjzdpj/external/bazel_tools/tools/cpp/lib_cc_configure.bzl:115:5:
Auto-Configuration Warning: 'BAZEL_VC' is not set, start looking for the latest Visual C++ installed.
DEBUG: C:/users/hubris/_bazel_hubris/gcsjzdpj/external/bazel_tools/tools/cpp/lib_cc_configure.bzl:115:5:
Auto-Configuration Warning: Looking for VS%VERSION%COMNTOOLS environment variables, eg. VS140COMNTOOLS
DEBUG: C:/users/hubris/_bazel_hubris/gcsjzdpj/external/bazel_tools/tools/cpp/lib_cc_configure.bzl:115:5:
Auto-Configuration Warning: Visual C++ build tools found at D:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\
INFO: SHA256 (https://codeload.github.com/google/protobuf/zip/106ffc04be1abf3ff3399f54ccf149815b287dd9) = 4d127e1c3608803bc845cc4b3c8c42a6a16263b6ddcb932b53049e7ec8101398
ERROR: D:/devel/gopath/src/github.com/prysmaticlabs/prysm/beacon-chain/BUILD.bazel:5:1: no such package '@com_github_urfave_cli//': no such package '@bazel_gazelle_go_repository_tools//': no such package '@com_github_pelletier_go_toml//': Traceback (most recent call last):
        File "C:/users/hubris/_bazel_hubris/gcsjzdpj/external/io_bazel_rules_go/go/private/tools/overlay_repository.bzl", line 67
                _apply_overlay(ctx, overlay)
        File "C:/users/hubris/_bazel_hubris/gcsjzdpj/external/io_bazel_rules_go/go/private/tools/overlay_repository.bzl", line 95, in _apply_overlay
                _check_execute(ctx, ["cp", src_path, dst_rel], ("fa...))
        File "C:/users/hubris/_bazel_hubris/gcsjzdpj/external/io_bazel_rules_go/go/private/tools/overlay_repository.bzl", line 100, in _check_execute
                fail((((message + "\n") + res.stdout)...))
failed to copy file from C:/users/hubris/_bazel_hubris/gcsjzdpj/external/io_bazel_rules_go/third_party/com_github_pelletier_go_toml/cmd/tomljson/BUILD.bazel.in
java.io.IOException: ERROR: src/main/native/windows/processes-jni.cc(239): CreateProcessW("cp" C:/users/hubris/_bazel_hubris/gcsjzdpj/external/io_bazel_rules_go/third_party/com_github_pelletier_go_toml/cmd/tomljson/BUILD.bazel.in cmd/tomljson/BUILD.bazel): The system cannot find the file specified.
 and referenced by '//beacon-chain:go_default_library'
ERROR: Analysis of target '//beacon-chain:beacon-chain' failed; build aborted: no such package '@com_github_urfave_cli//': no such package '@bazel_gazelle_go_repository_tools//': no such package '@com_github_pelletier_go_toml//': Traceback (most recent call last):
        File "C:/users/hubris/_bazel_hubris/gcsjzdpj/external/io_bazel_rules_go/go/private/tools/overlay_repository.bzl", line 67
                _apply_overlay(ctx, overlay)
        File "C:/users/hubris/_bazel_hubris/gcsjzdpj/external/io_bazel_rules_go/go/private/tools/overlay_repository.bzl", line 95, in _apply_overlay
                _check_execute(ctx, ["cp", src_path, dst_rel], ("fa...))
        File "C:/users/hubris/_bazel_hubris/gcsjzdpj/external/io_bazel_rules_go/go/private/tools/overlay_repository.bzl", line 100, in _check_execute
                fail((((message + "\n") + res.stdout)...))
failed to copy file from C:/users/hubris/_bazel_hubris/gcsjzdpj/external/io_bazel_rules_go/third_party/com_github_pelletier_go_toml/cmd/tomljson/BUILD.bazel.in
java.io.IOException: ERROR: src/main/native/windows/processes-jni.cc(239): CreateProcessW("cp" C:/users/hubris/_bazel_hubris/gcsjzdpj/external/io_bazel_rules_go/third_party/com_github_pelletier_go_toml/cmd/tomljson/BUILD.bazel.in cmd/tomljson/BUILD.bazel): The system cannot find the file specified.
INFO: Elapsed time: 4.093s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (7 packages loaded)
```

hubris replied:
These errors are weird:
``no such package '@com_github_urfave_cli//': no such package '@bazel_gazelle_go_repository_tools//': no such package '@com_github_pelletier_go_toml//': ``

prestonvanloon replied:
OK, maybe we don't have build on windows support. I believe no one is working on windows so this might have been broken for some time. 

Let's go ahead with win64 as you have it and assume that it will work when the other issues are resolved.

Where you able to find or build a binary for mac?

hubris replied:
@prestonvanloon I didn't find a binary for mac and I don't have one to build. If you can provide me a binary I could try to integrate it.

prestonvanloon replied:
The homebrew uses the tar.gz. 
See: https://github.com/ethereum/homebrew-ethereum/blob/master/solidity.rb

prestonvanloon replied:
Although it uses cmake to build it. Maybe we need to submit a feature request to solidity to start providing the mac binaries.

hubris replied:
@prestonvanloon  There is one already: https://github.com/ethereum/solidity/issues/3168

prestonvanloon commented on contracts/rules/solcompiler.bzl:
```diff
@@ -49,6 +68,8 @@ BINFILE=$OUTDIR/$CONTRACTNAME.bin
 
 def _solc_fetch_impl(ctx):
   #Detect host (same code as go_download_sdk)
+  print("FDSF")
```
What's this?

prestonvanloon commented:
I tried building a binary on Mac but it looks like it requires some dynamically linked libraries? ugh.
It was specifically failing for z3 not being present. 

https://travis-ci.org/prysmaticlabs/prysm/builds/426989709

Any ideas here? I'd love to get this merged, but I'm not sure how to make a MacOS distributable solc... yet. 

Filed a feature request here: https://github.com/ethereum/solidity/issues/4942

hubris commented:
@prestonvanloon I think it should be possible to pack the needed librarylike it is done on windows. If you do a ``otool -L solc`` we could get the list of required libraries, pack them into a zip and we should be able to launch the binary.
 I unfortunately don't have a mac to do this but I think it should work and may be better that static binary according to this https://developer.apple.com/library/archive/qa/qa1118/_index.html:

> Apple does not support statically linked binaries on Mac OS X.

prestonvanloon commented on contracts/rules/solcompiler.bzl:
```diff
@@ -0,0 +1,114 @@
+
+#Generate abi file
+$SOLCOMPILER --overwrite --bin --abi -o $OUTDIR $1
+CONTRACTNAME=$($SOLCOMPILER  --abi  $1  | sed -n 's,^=.*:\(.*\) .*=$,\\1,p')
```
on mac, this value is empty. Not sure why yet...

hubris replied:
@prestonvanloon Hi, I managed to install a macos VM (didn't know it aws possible), I will try to fix the problem.
